src/services/data-management/gateway-persistence.service.ts

```diff
@@ -0,0 +1,124 @@
+import { ChirpstackMQTTConnectionStateMessageDto } from "@dto/chirpstack/chirpstack-mqtt-message.dto";
+import { RawGatewayStateDto } from "@dto/kafka/raw-gateway-state.dto";
+import { GatewayStatusHistory } from "@entities/gateway-status-history.entity";
+import { KafkaTopic } from "@enum/kafka-topic.enum";
+import { subtractDays, subtractHours } from "@helpers/date.helper";
+import { Injectable, Logger } from "@nestjs/common";
+import { InjectRepository } from "@nestjs/typeorm";
+import { AbstractKafkaConsumer } from "@services/kafka/kafka.abstract.consumer";
+import { CombinedSubscribeTo } from "@services/kafka/kafka.decorator";
+import { KafkaPayload } from "@services/kafka/kafka.message";
+import { LessThan, MoreThan, Repository } from "typeorm";
+
+@Injectable()
+export class GatewayPersistenceService extends AbstractKafkaConsumer {
+    private readonly gatewayStatusSavedDays = 30;
+    /**
+     * Limit how many messages can be stored for each hour
+     */
+    private readonly maxStatusMessagesPerHour = 10;
+
+    constructor(
+        @InjectRepository(GatewayStatusHistory)
+        private gatewayStatusHistoryRepository: Repository<GatewayStatusHistory>
+    ) {
+        super();
+    }
+
+    private readonly logger = new Logger(GatewayPersistenceService.name);
+
+    protected registerTopic(): void {
+        this.addTopic(KafkaTopic.RAW_GATEWAY_STATE, "GatewayPersistence");
+    }
+
+    // Listen to Kafka event
+    @CombinedSubscribeTo(KafkaTopic.RAW_GATEWAY_STATE, "GatewayPersistence")
+    async rawRequestListener(payload: KafkaPayload): Promise<void> {
+        this.logger.debug(`RAW_GATEWAY_STATE: '${JSON.stringify(payload)}'`);
+        const dto = payload.body as RawGatewayStateDto;
+        const messageState = (dto.rawPayload as unknown) as ChirpstackMQTTConnectionStateMessageDto;
+
+        const statusHistory = this.mapDtoToEntity(dto, messageState);
+        await this.gatewayStatusHistoryRepository.save(statusHistory);
+
+		// Clean up old statuses
+        await this.deleteStatusHistoriesSinceLastHour(
+            statusHistory.timestamp,
+            dto.gatewayId
+        );
+        await this.deleteOldStatusHistories(dto.gatewayId);
+    }
+
+    private mapDtoToEntity(
+        dto: RawGatewayStateDto,
+        messageState: ChirpstackMQTTConnectionStateMessageDto
+    ) {
+        const statusHistory = new GatewayStatusHistory();
+        statusHistory.mac = dto.gatewayId;
+        statusHistory.timestamp = dto.unixTimestamp
```

Comment:
Is this going to be an issue with UTC?

Reply:
Currently, the timestamp is set and passed from [ReceiveDataService.buildMessage()](https://github.com/OS2iot/OS2IoT-backend/pull/170/files#diff-09adc96f1df326bcc3a378c68f40e9bc0bdcb3dcffd741262e1a14feb1748f37R58). The value is `new Date().valueOf()` which is always the unix time in ms.
When we store a Date object, they are converted to UTC time.